lib/spack/llnl/util/link_tree.py

```diff
@@ -150,15 +151,27 @@ def visit_file(self, root: str, rel_path: str, depth: int) -> None:
                 )
             )
         elif proj_rel_path in self.files:
-            # In some cases we can resolve file-file conflicts
+            # Resolve file-file conflicts when two files are identical: this deals with cases of
+            # of symlink to symlink, symlink to regular file and hardlinked files, where in
+            # principle the file we pick is arbitrary. However, for environment views that copy,
+            # we want the file to be copied, not the symlink, so we pick the underlying file, which
+            # should be the current, assuming proper ordering.
+            # NOTE: in principle filecmp could be used too as a more generic solution, but that is
+            # relatively slow, and so far this visitor has only used file metadata, not contents.
             src_a_root, src_a_relpath = self.files[proj_rel_path]
-            self.file_conflicts.append(
-                MergeConflict(
-                    dst=proj_rel_path,
-                    src_a=os.path.join(src_a_root, src_a_relpath),
-                    src_b=os.path.join(root, rel_path),
+
+            src_a = os.path.join(src_a_root, src_a_relpath)
+            src_b = os.path.join(root, rel_path)
+
+            if os.path.samefile(src_a, src_b):
+                # Delete and re-insert to retain correct self.files ordering.
```

Comment:
Thanks, I thought that was an issue of the past.

Reply:
`link_tree` has existed for a while and used `dict`, and didn't depend on ordering. So being careful here seems to be introducing new API guarantees. I'm inclined to say that:

* For now don't explicitly `del`, or
* Add a comment here about how we want to improve `link_tree`'s behavior to do an ordered traversal and we are leaning on the (implementation-specific) details of python dictionaries

Reply:
The order is relied on in `SimpleFilesystemView` for about two years. This PR does not introduce new behavior or API. The `del` is necessary, if you remove it a test fails.

If you are concerned about this detail, can't you deal with it in a separate PR? In principle this PR can be backported as a bugfix, I don't want to lay out a plan of future improvements in comments, cause I don't think it needs to be, I'd rather leave that to you as separate PR.

Reply:
> The del is necessary, if you remove it a test fails.

Are you saying tests other than the ones added in this PR would fail if this `del`+reinsert logic was just an override?

Or that the logic in general would fail if this overwrite-without-delete occurred (because perhaps the dirs need to be made before the files in them)?

I'm not sure how either would be a problem: if it was ok to have `src_a` at this point, it seems like it's always OK to have `src_b` inserted at the same time.

Reply:
Files are later grouped by prefix. Group by assumes consecutive files from same prefix.

Reply:
OK, how about we just augment the comment to point to this function (`_source_merge_visitor_to_merge_map`). Ideally we'd also update the docstring for SourceMergeVisitor to make this official. IMO this is also a good cautionary tale against reaching into the members of objects.

I want it to be clear we depend on preservation of order as an internal implementation detail. Ideally we could have a `_replace_file_item` function that codifies this internal constraint.